### PR TITLE
DM-14191: Merge butler configuration files with defaults

### PR DESCRIPTION
**Pull request by timj**
* Also rewrite `doImport`

### CODE REVIEW

File: python/lsst/daf/butler/core/butlerConfig.py
Comment (pschella) [inline python/lsst/daf/butler/core/butlerConfig.py]: Better to make `validate` a public class method instead?

File: python/lsst/daf/butler/core/datastore.py
Comment (pschella) [inline python/lsst/daf/butler/core/datastore.py:37]: Isn't this check now duplicated in `Butler(Config)`?

File: python/lsst/daf/butler/core/registry.py
Comment (pschella) [inline python/lsst/daf/butler/core/registry.py]: If `config` now is a `RegistryConfig` doesn't it have just the `registry` subset?

Comment (pschella) [reply python/lsst/daf/butler/core/registry.py]: Ah, I see, because we want the full Butler config for it. Maybe then explicitly specify that?

File: python/lsst/daf/butler/core/butlerConfig.py
Comment (TallJimbo) [inline python/lsst/daf/butler/core/butlerConfig.py]: `utils.getPackageDir`?

File: python/lsst/daf/butler/core/butlerConfig.py
Comment (TallJimbo) [inline python/lsst/daf/butler/core/butlerConfig.py]: Might want to rename this argument to something more descriptive.

File: config/datastores/posixDatastore.yaml
Comment (TallJimbo) [inline config/datastores/posixDatastore.yaml]: I'm a bit surprised to not see the default mapping from StorageClass to Formatter here.  Those are specific for PosixDatastore, right?

I don't actually care if they go here or elsewhere, but we need an easy way to construct a Butler that knows about those defaults.

File: python/lsst/daf/butler/core/butlerConfig.py
Comment (TallJimbo) [inline python/lsst/daf/butler/core/butlerConfig.py]: Remove or use logging.

File: python/lsst/daf/butler/core/butlerConfig.py
Comment (TallJimbo) [inline python/lsst/daf/butler/core/butlerConfig.py]: Should we be using `RegistryConfig` here?  ("No" is a perfectly good answer; I'm just curious why not.)

**Review by TallJimbo — APPROVED** [review]
Looks good; mostly very minor comments.  I would like to make sure we figure out where to put the StorageClass->Formatter mapping defaults so those don't have to be repeated in each repository (and that raises the question of which config values, if any, should always go in the repo config instead of being pulled from the defaults).

**Comment by pschella** [comment]
Since they are all defaults we may want to rename `default_schema.yaml` to something else.

File: python/lsst/daf/butler/core/butlerConfig.py
Comment (pschella) [inline python/lsst/daf/butler/core/butlerConfig.py]: I still wonder if having just one level of overrides is future proof enough. As opposed to a (colon-separated) `$DAF_BUTLER_CONFIG_PATH`.

Comment (timj) [reply python/lsst/daf/butler/core/butlerConfig.py]: I can do that. I thought we agreed that once we have the merging code in one place we can easily add it but since I'm there I'll add this other variable as well.

File: python/lsst/daf/butler/core/datastore.py
Comment (timj) [reply python/lsst/daf/butler/core/datastore.py:37]: ButlerConfig doesn't inherit from ConfigSubset so no. I could move the validation code in `ConfigSubset.__init__()` to `Config.__init__()` and always have it as an option if you set `requiredKeys`.

File: python/lsst/daf/butler/core/butlerConfig.py
Comment (timj) [reply python/lsst/daf/butler/core/butlerConfig.py]: See my comment below on validation of `ConfigSubset` but would you prefer class attributes `Config.requiredKeys` and a public method in `Config.validate` which used `requiredKeys` but could be overridden? Why a class method and not an instance method?

File: python/lsst/daf/butler/core/registry.py
Comment (timj) [reply python/lsst/daf/butler/core/registry.py]: Right. I think it's better for RegistryConfig to only have registry stuff and SchemaConfig to only have schema stuff. I'll change that.

File: python/lsst/daf/butler/core/butlerConfig.py
Comment (timj) [reply python/lsst/daf/butler/core/butlerConfig.py]: I called it that because that's what it was called in `Config` (which came from `daf_persistence`). I'll have a think.

File: config/datastores/posixDatastore.yaml
Comment (timj) [reply config/datastores/posixDatastore.yaml]: I wasn't entirely sure what defaults would mean since the formatters really do lock in the file format, but you are right that FITS is where we are so our default for "how to write an Exposure" should be in this file.

File: python/lsst/daf/butler/core/butlerConfig.py
Comment (timj) [reply python/lsst/daf/butler/core/butlerConfig.py]: Yes. Sorry. Should have deleted it.

File: python/lsst/daf/butler/core/butlerConfig.py
Comment (timj) [reply python/lsst/daf/butler/core/butlerConfig.py]: Probably. Before I added `ConfigSubset` there wasn't any difference. Now there is so I'll revisit that.

Comment (timj) [reply python/lsst/daf/butler/core/butlerConfig.py]: Although, these configs have to fit in the global structure so should not have the `schema` stripped. The advantage of using `SchemaConfig` is you get the validation so I have to see if I can read the subset and bind it to the larger config hierarchy.

File: python/lsst/daf/butler/core/butlerConfig.py
Comment (timj) [reply python/lsst/daf/butler/core/butlerConfig.py]: I wondered about that. `utils` would work fine but I do wonder about adding LSST dependencies for no real gain. We only use `utils` in tests and I would argue that we don't need that dependency. That leaves only `lsst.log` and `sphgeom` stopping us from putting this package directly on pypi for everyone to use without needing any non-standard dependencies. (I'm thinking about world domination here...)

**Comment by timj** [comment]
Regarding `default_schema.yaml` I'm happy to delete the `registry` directory and call the file `registry_schema.yaml`.

File: python/lsst/daf/butler/core/butlerConfig.py
Comment (timj) [reply python/lsst/daf/butler/core/butlerConfig.py]: Do you envisage the `$DAF_BUTLER_CONFIG_PATH` containing a set of search paths that look identical to the on in `daf_butler/config` (so we look for `storageClasses.yaml`, etc)? And that the defaults should be filled in by reading butler configs first, then the path in reverse order?

Comment (TallJimbo) [reply python/lsst/daf/butler/core/butlerConfig.py]: I have some of the same thoughts about world domination (see RFC-430 and RFC-460 vis a vis getting that sphgeom dependency out on pypi).  But I think using `DAF_BUTLER_DIR` here just avoids an explicit dependency on `utils` via an implicit dependency on `EUPS` and perhaps `sconsUtils`, and I think the `utils` version may be slightly easier to rip out and replace with something unencumbered later.

Comment (TallJimbo) [reply python/lsst/daf/butler/core/butlerConfig.py]: I think we should hold off on this until we have a use case for it.  I tend to think that having extra paths that aren't connected to either a repo or a software version could do more harm than good.

Comment (timj) [reply python/lsst/daf/butler/core/butlerConfig.py]: I implemented this last night... How about I push the change this morning as a new commit and you can take a look.

File: config/datastores/posixDatastore.yaml
Comment (TallJimbo) [reply config/datastores/posixDatastore.yaml]: A lot of those defaults currently live in the `gen3` module of `ci_hsc` (on the `gen3-middleware` branch).  I'd love to get those moved out of Python code into YAML.  I'd also be happy to take care of making a ci_hsc branch compatible with this ticket once you're done with the daf_butler changes.

File: python/lsst/daf/butler/core/butlerConfig.py
Comment (ktlim) [inline python/lsst/daf/butler/core/butlerConfig.py]: Doesn't this apply absolute paths over and over?  Maybe that's not a problem.

Comment (timj) [reply python/lsst/daf/butler/core/butlerConfig.py]: Yes. It does. Hadn't thought of that. I don't think it's a problem as such, but it will slow things down.

File: python/lsst/daf/butler/core/datastore.py
Comment (pschella) [reply python/lsst/daf/butler/core/datastore.py:37]: That sounds good.

File: python/lsst/daf/butler/core/butlerConfig.py
Comment (pschella) [inline python/lsst/daf/butler/core/butlerConfig.py]: Is this opposite behavior to shell `$PATH`? If so why?

File: python/lsst/daf/butler/core/butlerConfig.py
Comment (pschella) [inline python/lsst/daf/butler/core/butlerConfig.py]: supplied butler.

File: python/lsst/daf/butler/core/butlerConfig.py
Comment (pschella) [inline python/lsst/daf/butler/core/butlerConfig.py]: Why not simply loop over all components and have those subclasses determine if they are special?

Comment (pschella) [reply python/lsst/daf/butler/core/butlerConfig.py]: E.g. by having a `component` member.

File: python/lsst/daf/butler/core/butlerConfig.py
Comment (pschella) [inline python/lsst/daf/butler/core/butlerConfig.py]: Perhaps make this a class member?

File: python/lsst/daf/butler/core/butlerConfig.py
Comment (pschella) [reply python/lsst/daf/butler/core/butlerConfig.py]: Why not simply invert the loop order?

File: python/lsst/daf/butler/core/config.py
Comment (pschella) [inline python/lsst/daf/butler/core/config.py]: `ppprint`?

File: python/lsst/daf/butler/core/config.py
Comment (pschella) [inline python/lsst/daf/butler/core/config.py]: What if `other` is a `ConfigSubset`?

File: tests/test_dataUnit.py
Comment (pschella) [inline tests/test_dataUnit.py]: Do we still need to specify the `schemaFile`? Or does calling `SchemaConfig()` load the default file?

File: tests/test_instrument.py
Comment (pschella) [inline tests/test_instrument.py:52]: Should we have a default file for `Butler` too perhaps?

Comment (pschella) [reply tests/test_instrument.py:52]: If only for users to copy. But it might be useful to do `butler = Butler()` and simply get something working with a local `SqliteRegistry` and `PosixDatastore` in a default local subdirectory.

Comment (pschella) [reply tests/test_instrument.py:52]: Perhaps even more useful if that then can write the default `butler.yaml` config in it (something like `git init .`) that you can subsequently modify.

Comment (pschella) [reply tests/test_instrument.py:52]: But maybe the default constructor is not the best for that. And instead you want an explicit class method e.g. `butler = Butler.makeDefault()` (with an optional path).

File: tests/test_posixDatastoreFits.py
Comment (pschella) [inline tests/test_posixDatastoreFits.py:138]: Are we now assuming that `Datastore` and `Registry` always have the same `root`? What if (for a particular datastore) there is no such thing?

File: python/lsst/daf/butler/core/config.py
Comment (TallJimbo) [inline python/lsst/daf/butler/core/config.py]: I think `pdb` requires parentheses after `print` these days.

File: python/lsst/daf/butler/core/config.py
Comment (TallJimbo) [inline python/lsst/daf/butler/core/config.py]: Missing the end of this sentence, I think.

File: python/lsst/daf/butler/core/config.py
Comment (TallJimbo) [inline python/lsst/daf/butler/core/config.py:390]: Are these relative to the top level or the component?  (should be documented)

File: python/lsst/daf/butler/core/butlerConfig.py
Comment (TallJimbo) [inline python/lsst/daf/butler/core/butlerConfig.py]: I do think we'll need to defer validation until the full config has been loaded and merged.

File: python/lsst/daf/butler/core/utils.py
Comment (TallJimbo) [inline python/lsst/daf/butler/core/utils.py]: I think it might be best to catch and re-raise any of these as `ImportError`; it's much more likely that the caller will want to catch anything (without having a too-general `except Exception` block) than try to distinguish between them.

File: python/lsst/daf/butler/registries/sqlRegistry.py
Comment (TallJimbo) [inline python/lsst/daf/butler/registries/sqlRegistry.py:60]: Do we not actually have a default for `SqlRegistry` configuration?

**Review by TallJimbo — APPROVED** [review]
Looks good.  I've made some minor comments, but like @pschella, from this point I'm happy for you to merge when you think it's in good shape.
If you do want to merge before moving the StorageClass->Formatter lookup defaults to a non-test config file, I'll go ahead and add those on DM-14378 (there are a number of additional mappings I want to move to daf_butler from ci_hsc anyway, once I have a place to put them in daf_butler).

File: python/lsst/daf/butler/core/utils.py
Comment (timj) [reply python/lsst/daf/butler/core/utils.py]: `ModuleNotFoundError` is a subclass of `ImportError`. I was pleased to discover that the simplification of `doImport` is only possible because we are using python 3.6 (previous versions did not have `ModuleNotFoundError`). I was trying to keep the interface looking like `import`. I suppose I could convert `AttributeError` to `ImportError` (although I thought I had managed to get `import` to throw that at one point, maybe via `from`).

Comment (TallJimbo) [reply python/lsst/daf/butler/core/utils.py]: Yeah, using a hierarchy of exceptions is the best scenario.  If you can convert the `AttributeError` to `ImportError`, I do think that'd be best, even if that is a bit unlike `import` itself.

File: tests/test_instrument.py
Comment (TallJimbo) [reply tests/test_instrument.py:52]: I think I'd go with an explicit class method with a required path argument (which could accept `"."`, I suppose).

**Comment by TallJimbo** [comment]
> Regarding default_schema.yaml I'm happy to delete the registry directory and call the file registry_schema.yaml.

Should we just call it `schema.yaml`, so overrides can use the same filename?   (Overrides should be rare, but probably permitted in the same way StorareClass overrides are).

**Comment by timj** [comment]
Everything can be overridden using `$DAF_BUTLER_CONFIG_PATH` if they use the same file name as is in the butler config directory. I'm okay with `schema.yaml`.

File: python/lsst/daf/butler/core/butlerConfig.py
Comment (timj) [reply python/lsst/daf/butler/core/butlerConfig.py]: It's not meant to be the opposite order. In `$PATH` things early in the path take precedence over things later in the path, so I thought I had to reverse the order when reading in the defaults so that earlier ones in the path override defaults appearing late in the path.

File: python/lsst/daf/butler/core/butlerConfig.py
Comment (timj) [reply python/lsst/daf/butler/core/butlerConfig.py]: yes. since the configs read from the paths are subsets of the config I can loop over configs and then paths.

File: python/lsst/daf/butler/core/butlerConfig.py
Comment (timj) [reply python/lsst/daf/butler/core/butlerConfig.py]: Are you saying I should look through all the top level keys in the supplied butler config looking for `.cls` members, then ask the imported classes for their config files and associated config classes? (the latter they don't know but they could).

File: python/lsst/daf/butler/core/config.py
Comment (timj) [reply python/lsst/daf/butler/core/config.py]: I didn't write the method...

File: python/lsst/daf/butler/core/config.py
Comment (timj) [reply python/lsst/daf/butler/core/config.py]: What am I missing?
```
>>> from lsst.daf.butler import ConfigSubset
>>> x = ConfigSubset({"A": 5, "b": 6})
>>> x
{'A': 5, 'b': 6}
>>> y = ConfigSubset(x)
>>> y
{'A': 5, 'b': 6}
>>> x["c"] = 3
>>> x
{'A': 5, 'b': 6, 'c': 3}
>>> y
{'A': 5, 'b': 6}
```

File: tests/test_dataUnit.py
Comment (timj) [reply tests/test_dataUnit.py]: At the moment, the only defaulting is in `ButlerConfig`. I can see how each subclass doing its own defaulting and getting the search path from a shared routine might be helpful (and more obvious).

File: tests/test_posixDatastoreFits.py
Comment (timj) [reply tests/test_posixDatastoreFits.py:138]: I'm not sure I understand. This is a `DatastoreConfig` so by definition it's a datastore config not a registry config. Or are you asking a different question?

File: python/lsst/daf/butler/core/config.py
Comment (timj) [reply python/lsst/daf/butler/core/config.py:390]: Relative to the Config root of this config (which does not include the component name).

File: python/lsst/daf/butler/core/config.py
Comment (TallJimbo) [inline python/lsst/daf/butler/core/config.py:407]: Does this mean we merge the wrong `externalConfig` values when these conditions are not met, or does that mean we're assuming that `externalConfig` already corresponds to just the stuff under this component?

File: python/lsst/daf/butler/core/config.py
Comment (TallJimbo) [inline python/lsst/daf/butler/core/config.py]: Orphaned sentence.

**Review by TallJimbo — APPROVED** [review]
Looks good; one question and one doc bug.

Also, could you rename `registry_schema.yaml` to just `schema.yaml`?

File: python/lsst/daf/butler/core/config.py
Comment (timj) [reply python/lsst/daf/butler/core/config.py:407]: It means that if you give a config with a root of "datastore" and the config knows it's a datastore then your config will take the contents of the "datastore" part of the tree. If there is no "datastore" in the tree it assumes it's all meant for this config (and will then check for mandatory keys). ie you can give it a datastore.yaml that doesn't have the "datastore" part.

**Pull request merged**
